specs/BUNDLE_SPEC.md

```diff
@@ -0,0 +1,320 @@
+  "schemaVersion": 2,
+  "artifactType": "application/vnd.dev.sigstore.bundle.v0.3+json",
+  "config": {
+    "mediaType": "application/vnd.oci.empty.v1+json",
+    "digest": "sha256:44136fa3...",
+    "size": 2
+  },
+  "layers": [
+    {
+      "digest": "sha256:cafed00d...",
+      "mediaType": "application/vnd.dev.sigstore.bundle.v0.3+json",
+      "size": 4971
+    }
+  ],
+  "subject": {
+    "digest": "sha256:c00010ff...",
+    "mediaType": "application/vnd.oci.image.index.v1+json"
+   }
+}
+```
+
+The final step is to use the `digest` from the first of the `layers` to retrieve
+the bundle blob:
+
+```
+GET /v2/foo/blobs/uploads/?digest=cafed00d...
+```
+
+```
+{
+  "mediaType": "application/vnd.dev.sigstore.bundle.v0.3+json",
+  "verificationMaterial": {...},
+  "messageSignature": {...}
+}
+```
+
+## Annotations
+
+For any given image, there may be any number of attached attestation bundles.
+When there are multiple Sigstore bundles associated with an image it may be
+difficult to identify which artifact is which in the image index:
+
+```json
+{
+  "mediaType": "application/vnd.oci.image.index.v1+json",
+  "schemaVersion": 2,
+  "manifests": [
+    {
+      "artifactType": "application/vnd.dev.sigstore.bundle.v0.3+json",
+      "digest": "sha256:facefeed",
+      "mediaType": "application/vnd.oci.image.manifest.v1+json"
+    },
+    {
+      "artifactType": "application/vnd.dev.sigstore.bundle.v0.3+json",
+      "digest": "sha256:d0d0caca",
+      "mediaType": "application/vnd.oci.image.manifest.v1+json"
+    }
+  ]
+}
+```
+
+To help disambiguate attestations, clients may add annotations to the items in
+the `manifests` list which indicate what is contained within each bundle and
+when it was created:
+
+- `dev.sigstore.bundle.content` - Must be one "message-signature" or
+  "dsse-envelope" and should match the type of content embedded in the Sigstore
+  bundle.
+- `dev.sigstore.bundle.predicateType` - When the bundle contains a DSSE-wrapped
+  in-toto statement, the statement's predicate can be reflected here.
+- `dev.sigstore.bundle.signer` - Identity of the application which generated the
+  attestation bundle.
+- `org.opencontainers.image.created` - Date and time when the attestation bundle
+  was created, conforming to
+  [RFC 3339](https://tools.ietf.org/html/rfc3339#section-5.6) (this is one of
+  the pre-defined annotation keys identified in the
+  [OCI spec](https://github.com/opencontainers/image-spec/blob/main/annotations.md#pre-defined-annotation-keys)).
+
+These annotations should be included as part of the bundle manifest:
+
+```json
+{
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "schemaVersion": 2,
+  "artifactType": "application/vnd.dev.sigstore.bundle.v0.3+json",
+  "annotations": {
+    "dev.sigstore.bundle.content": "dsse-envelope",
+    "dev.sigstore.bundle.predicateType": "https://slsa.dev/provenance/v1",
+    "dev.sigstore.bundle.signer": "cosign/v2.2.3 (darwin; arm64)",
```

Comment:
For the signer details, the concept I was thinking of would include details used by certificate-identity, certificate-oidc-issuer, etc, so that a `cosign verify` command could quickly find the matching signature. That could end up being more than one annotation.

Reply:
I'd worry that would get used to build a verification policy rather than that policy be provided from the caller. I also don't want the annotations to become a duplicate of what's in the envelope/certificate.

Reply:
There may be multiple signatures uploaded for a given digest (hundreds if someone has reproducible builds or runs something daily). Ideally, each of those wouldn't need to be pulled separately to find the one matching the verification policy the user is asking for. This would be an API efficiency, and security would still be provided by validating the envelope/certificate itself.

Reply:
If we have multiple attestations for the same digest (some recurring, reproducible build) it's likely that the predicate, cert identity, OIDC issuer will be the same for all of them -- surfacing that info as an annotation probably isn't gonna be that useful.

I was chatting w/ @codysoyland who is working on bundle support in the policy-controller and his primary requirement was that the predicate-type be available (as this is a required part of the [policy definition](https://github.com/actions/attest-build-provenance/pull/33)).

At this point, I'm inclined to trim the specified annotation list to just the `content`, `predicateType` and the `created` values. Leaving open the option to add more annotations in the future as specific use cases arise.